### PR TITLE
Add set of APIs for managing scenarios and scenario results.

### PR DESCRIPTION
**Pull request by runforyourlives**
Adds ability to create, delete, list, and retrieve scenarios.
Adds ability to update a scenario's result.

Not completed yet:
 - API to retrieve a scenario's results given a scenario ID.

We expect to add more functionality when the scenario result fields become richer.

### CODE REVIEW

File: src/planscape/planning/tests.py [hunk 1]
```diff
@@ -363,3 +363,383 @@ def test_end_to_end(self):
...
+            content_type="application/json")
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(Scenario.objects.count(), 1)
+        self.assertEqual(ScenarioResult.objects.count(), 1)
+
+    def test_create_scenario_not_logged_in(self):
+        response = self.client.post(
+            reverse('planning:create_scenario'),
+            {'plan_id': self.plan.pk,
+             'configuration': '{}',
```
Comment (sachiyuan) [inline src/planscape/planning/tests.py:416]: Once we have configuration validation we'll want a test_configuration dictionary with basic values to pass in here. Might be worth adding a todo or making an empty one for now to pass in just as a reminder

File: src/planscape/planning/views.py [hunk 2]
```diff
@@ -182,3 +181,196 @@ def list_plans(request: HttpRequest) -> HttpResponse:
...
+#TODO
+#def list_results_for_scenario(request: HttpRequest) -> HttpResponse:
+
+# TODO: add more things to update other than state
+# Can update a state only such that pending -> running | failure, and running -> success | failure
+# Throws an error if no scenario owned by the user can be found with the desired ID.
+def update_scenario_result(request: HttpRequest) -> HttpResponse:
+    try:
+        # Check that the user is logged in.
+        user = _get_user(request)
```
Comment (sachiyuan) [inline src/planscape/planning/views.py:274]: Would there be a user associated with this request? (Aka is this a request that a user would make? Or is this called completely from the backend?)

File: src/planscape/planning/views.py [hunk 3]
```diff
@@ -182,3 +181,196 @@ def list_plans(request: HttpRequest) -> HttpResponse:
...
+            scenario_ids = scenario_id_str
+        else:
+            raise ValueError("scenario_id must be an int or a list of ints.")
+
+        # Get the scenarios matching the provided IDs and the logged-in user.
+        scenarios = Scenario.objects.filter(pk__in=scenario_ids).filter(planning_area__user=user.pk)
+        # This automatically deletes ScenarioResult entries for the deleted Scenarios.
+        scenarios.delete()
+
+        # We still report that the full set of plan IDs requested were deleted,
```
Comment (sachiyuan) [inline src/planscape/planning/views.py:366]: Can you clarify this comment a bit? Is it saying even if the user is wrong, it still reports the objects were deleted? 

(Also "plan ID's" -> "scenario ID's"?)

File: src/planscape/planning/tests.py (hunk 1 above)
Comment (runforyourlives) [reply src/planscape/planning/tests.py:416]: Yup - the TODO's on line 441.  I do have the empty config here in the tests, but if you already have some sort of structure to suggest, I'll use it here.

Comment (sachiyuan) [reply src/planscape/planning/tests.py:416]: So far it looks like an example one might look like:
{
    'est_cost': 0,
    'max_budget': 0,
    'max_road_distance', 0,
    'max_slope': 0,
    'priorities': ['priority1'],
    'weights': [0],
    'stand_size': 'Large',
   'excluded_areas': []
 }

Note for future: I believe 'max_budget' will be an optional config value

File: src/planscape/planning/views.py [hunk 4]
```diff
@@ -182,3 +181,196 @@ def list_plans(request: HttpRequest) -> HttpResponse:
...
+        # Check that the user is logged in.
+        user = _get_user(request)
+        if user is None:
+            raise ValueError("User must be logged in.")
+
+        body = json.loads(request.body)
+    
+        plan = get_object_or_404(user.planning_areas, id=body['plan_id'])
+        
+        configuration = body.get('configuration')
```
[the remark this answers is not in the document]
Comment (runforyourlives) [reply src/planscape/planning/views.py:240]: I ended up switching to the serializer approach and changing the plan_id input to planning_area.  It was a much smoother change than I anticipated.

File: src/planscape/planning/urls.py [hunk 5]
```diff
@@ -11,7 +13,11 @@
     path('list_plans/', list_plans, name='list_plans'),
 
     # Scenarios
-    # TODO
+    path('create_scenario/', create_scenario, name='create_scenario'),
```
[the remark this answers is not in the document]
Comment (runforyourlives) [reply src/planscape/planning/urls.py:16]: Agreed that this is a good idea, but will be left as a TODO for time considerations.

File: src/planscape/planning/views.py [hunk 6]
```diff
@@ -182,3 +184,196 @@ def list_plans(request: HttpRequest) -> HttpResponse:
...
+
+        return JsonResponse(
+            _serialize_scenario(scenario, scenario_result),
+            safe=False)
+    except Exception as e:
+        return HttpResponseBadRequest("Ill-formed request: " + str(e))
+
+# Requires a logged in user, a plan_id, and a configuration for the new scenario.
+# This creates both the scenario and a corresponding ScenarioResult entry for that scenario.
+# Note that tehe ScenarioResult is by design the default one; you cannot insert a new ScenarioResult with
```
[the remark this answers is not in the document]
Comment (runforyourlives) [reply src/planscape/planning/views.py:227]: Fixed.

File: src/planscape/planning/views.py (hunk 2 above)
Comment (runforyourlives) [reply src/planscape/planning/views.py:274]: Wow.  This is a really good catch, as the EP won't have a user login context.  Updated, but we'll have a TODO in order to lock down access to just the EP and tests.

File: src/planscape/planning/views.py (hunk 3 above)
Comment (runforyourlives) [reply src/planscape/planning/views.py:366]: Reworded.

**Review by sachiyuan — APPROVED** [review]
Lgtm

**Pull request merged**